src/driver.rs

```diff
@@ -53,9 +55,35 @@ impl Driver {
         &self,
         py: Python<'_>,
         cluster: &magnum::Cluster,
+        upgrade: bool,
     ) -> PyResult<()> {
         py.allow_threads(|| {
             get_runtime().block_on(async {
+                // NOTE(mnaser): The updated cloud provider resource uses a different set of
+                //               labels and annotations (from the Helm chart) than the legacy
+                //               ones which was created by manifests. We need to clean up the
+                //               legacy resources otherwise it will generate a conflict during
+                //               the upgrade.
+                //
+                //               https://github.com/vexxhost/magnum-cluster-api/issues/580
+                if upgrade {
+                    debug!("Detecting cluster upgrade, ensuring that the legacy resource set is deleted");
+
+                    let client = cluster.client().await?;
+                    let daemonsets = client.list_resources(
+                        Api::<DaemonSet>::namespaced(client.clone(), "kube-system"),
+                        ListParams::default().labels("k8s-app=openstack-cloud-controller-manager"),
```

Comment:
This arbitrarily deletes things all the time, no?  Wouldn't we want to make sure that we only delete the ones that using the old labels and keep the ones with the new labels (aka noop)?

In addition, we need to do this for the Cinder CSI and Manila CSI as well.

Reply:
I defined addons for cinder-csi and manila-csi.
Now the next step is to offload these CSIs from legacy CRS and create dedicated CRSs for them separately and use addons.

Reply:
@okozachenko1203 I think we are adding extra scope here, I am not opposed to doing this, however, the issue is currently upgrades are broken because of this.

We really need to ship an upgrade fix, so I think we should keep this narrowly focused to upgrades, and then we can do the follow up for Manila and Cinder.

Reply:
@mnaser got it